### ISSUE TITLE
[url_launcher] Unupdated url_launcher_android dependency will cause version 2.1.0 to not work correctly

### ISSUE THREAD

**Issue by gabrielaraujoz** [flutter/flutter#102996]
The current url_launcher pubspec.yaml (version 6.1.0) will not force the update of its url_launcher_android dependency from 6.0.13 to the newest version 6.0.16. 

The 6.0.16 version for url_launcher_android will effectively fix the problem targeted by the url_launcher update 2.1.0, but since it's not correctly indicated in url_launcher's pubspec, devs would have the manually update their pubspec.lock files to use the correct version.

The pictures below should illustrate it better. 

url_launcher changelog stating the fix for preventing common issues with providing invalid URL strings.
![Screen Shot 2022-05-03 at 18 23 26](https://user-images.githubusercontent.com/20102814/166568786-09d43af8-0184-4f06-a2d7-690b0b8aadd0.png)

url_launcher's pubspec.yaml
![Screen Shot 2022-05-03 at 18 19 26](https://user-images.githubusercontent.com/20102814/166568923-b7f7ee19-754a-408c-8029-e62a04c7d07e.png)

url_launcher_android's changelog - fix only applied in version 6.0.16
![Screen Shot 2022-05-03 at 18 18 01](https://user-images.githubusercontent.com/20102814/166569007-acf61361-eddf-4ede-9c50-c832f52958cb.png)

I'm opening a PR with this pubspec update and will leave it there for your appreciation. Thanks!

**Comment by stuartmorgan** [comment]
> The current url_launcher pubspec.yaml (version 6.1.0) will not force the update of its url_launcher_android dependency from 6.0.13 to the newest version 6.0.16.

This is intentional; `url_launcher` 6.1.0 is compatible with earlier versions of `url_launcher_android`.

> The 6.0.16 version for url_launcher_android will effectively fix the problem targeted by the url_launcher update 2.1.0

`url_launcher_android` 6.0.16 fixes an issue with false negatives in `canLaunch*` for certain http(s) URLs on Android. `url_launcher` 6.1.0 provides a new API surface that is harder to use incorrectly. The two are unrelated.

> but since it's not correctly indicated in url_launcher's pubspec, devs would have the manually update their pubspec.lock files

`dart pub upgrade`, with or without a target package, will upgrade dependencies transitively. Nobody needs to manually edit pubspec.lock files to get updates to transitive dependencies.

> to use the correct version.

There is no "correct" here. People who want updated version of their transitive dependencies need to update them, yes, but if they don't they are welcome to continue using older versions.

Closing as there is no bug here. The title's claim that the combination will not work correctly is not correct.

**Comment by gabrielaraujoz** [comment]
@stuartmorgan Actually it won't launch if we don't update the url_launcher_android lib to 2.0.16. Here's a minimal version showing the bug:

run flutter create project_name
import url_launcher
Replace main.dart with this snippet:
```
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart' as url_launcher;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or simply save your changes to "hot reload" in a Flutter IDE).
        // Notice that the counter didn't reset back to zero; the application
        // is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  bool result = true;
  final link = Uri.parse('https://www.linkedin.com/in/gabrielaraujoz');

  void _incrementCounter() async {
    if (await url_launcher.canLaunchUrl(link)) {
      result = await url_launcher.launchUrl(link);
    } else {
      print('url cannot be launched, stackTrace: ${StackTrace.current}');
      result = false;
    }
    setState(() {});
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: Text(widget.title),
      ),
      body: Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: Column(
          // Column is also a layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug painting" (press "p" in the console, choose the
          // "Toggle Debug Paint" action from the Flutter Inspector in Android
          // Studio, or the "Toggle Debug Paint" command in Visual Studio Code)
          // to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            const Text(
              'This button will launch the link',
            ),
            Text(
              result.toString(),
              style: Theme.of(context).textTheme.headline4,
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Launch link',
        child: const Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
```

You'll see it won't launch, the error being:

```
I/UrlLauncher(29601): component name for https://www.linkedin.com/in/gabrielaraujoz is null
I/UrlLauncher(29601): component name for https://flutter.dev is null
I/flutter (29601): url cannot be launched, stackTrace: #0      _MyHomePageState._incrementCounter
package:url_launcher_test/main.dart:59
I/flutter (29601): <asynchronous suspension>
```

Could you please try this out and see if it works for you in an Android Emulator/Device? If it does it might be some configuration I have in my machine. Thanks!

**Comment by gabrielaraujoz** [comment]
@stuartmorgan honest question here: should I open another issue with this info as well?

**Comment by gabrielaraujoz** [comment]
Ok @stuartmorgan , thanks for the review and input. Sorry for wasting your time, I really thought this was necessary since in my project (with url_launcher_android at 6.0.13) it wouldn't launch the link and only got to work with the update to 6.0.16 - maybe we should add some documentation on this or do you think it shouldn't be necessary?

**Comment by stuartmorgan** [comment]
> I really thought this was necessary since in my project (with url_launcher_android at 6.0.13) it wouldn't launch the link and only got to work with the update to 6.0.16

If "the link" is a link with a custom app handler, and "won't launch the link" means that `canLaunch*` returns false (which isn't the same as not actually being able to launch the link, since `launch*` will work in that scenario—it's a false negative in `canLaunch`) that's expected, because `url_launcher_android` 6.0.16 is what fixed that specific issue. As documented in the `url_launcher_android` release notes in your screenshot above.

> maybe we should add some documentation on this

It's not clear to me what you want to document (or where). The issue you had is that you wanted a fix from `url_launcher_android` 6.0.16 but were using 6.0.13, and the solution is simply to update your package dependencies to get 6.0.16. How to update dependencies is [already documented](https://dart.dev/tools/pub/cmd/pub-upgrade).
